client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesClientIT.java

```diff
@@ -495,4 +498,125 @@ public void testRollover() throws IOException {
+        assertThat(getAliasesResponse.getAliases().get("index3").size(), equalTo(0));
+    }
+
+    public void testGetAliasesNonExistentIndexOrAlias() throws IOException {
+
+        String alias = "alias";
+        String index = "index";
+
+        GetAliasesRequest getAliasesRequest1 = new GetAliasesRequest().indices(index);
+        ElasticsearchException exception = expectThrows(ElasticsearchException.class,
+                () -> execute(getAliasesRequest1, highLevelClient().indices()::getAlias, highLevelClient().indices()::getAliasAsync));
+
+        assertThat(exception.status(), equalTo(RestStatus.NOT_FOUND));
+        assertThat(exception.getMessage(), equalTo("Elasticsearch exception [type=index_not_found_exception, reason=no such index]"));
+
+        GetAliasesRequest getAliasesRequest2 = new GetAliasesRequest(alias);
+        exception = expectThrows(ElasticsearchException.class,
+                () -> execute(getAliasesRequest2, highLevelClient().indices()::getAlias, highLevelClient().indices()::getAliasAsync));
+
+        assertThat(exception.status(), equalTo(RestStatus.NOT_FOUND));
+        assertThat(exception.getMessage(), equalTo("Elasticsearch exception [type=exception, reason=alias [" + alias + "] missing]"));
+
+        createIndex(index, Settings.EMPTY);
+        client().performRequest(HttpPut.METHOD_NAME, index + "/_alias/" + alias);
+
+        GetAliasesRequest getAliasesRequest3 = new GetAliasesRequest().indices(index, "non_existent_index");
+        exception = expectThrows(ElasticsearchException.class,
+                () -> execute(getAliasesRequest3, highLevelClient().indices()::getAlias, highLevelClient().indices()::getAliasAsync));
+        assertThat(exception.status(), equalTo(RestStatus.NOT_FOUND));
+        assertThat(exception.getMessage(), equalTo("Elasticsearch exception [type=index_not_found_exception, reason=no such index]"));
+
+        GetAliasesRequest getAliasesRequest4 = new GetAliasesRequest().indices(index, "non_existent_index").aliases(alias);
+        exception = expectThrows(ElasticsearchException.class,
+                () -> execute(getAliasesRequest4, highLevelClient().indices()::getAlias, highLevelClient().indices()::getAliasAsync));
+        assertThat(exception.status(), equalTo(RestStatus.NOT_FOUND));
+        assertThat(exception.getMessage(), equalTo("Elasticsearch exception [type=index_not_found_exception, reason=no such index]"));
+
+        GetAliasesRequest getAliasesRequest5 = new GetAliasesRequest().indices(index).aliases(alias, "non_existent_alias");
+        exception = expectThrows(ElasticsearchException.class,
+                () -> execute(getAliasesRequest5, highLevelClient().indices()::getAlias, highLevelClient().indices()::getAliasAsync));
+        /*
+        {
+         "error": "alias [something] missing",
+         "status": 404,
+         "index": {
+           "aliases": {
+             "alias": {}
+           }
+         }
+        }
+        */
```

Comment:
That is actually what ES Rest layer is returning in this case... And the high level client cannot parse the exception body here. 
Is this the only action which wraps the response into the returned exception ? Should I try to parse the response here ?

Reply:
this is a funny response, as the error may be mixed up with an actual response that holds aliases. We have similar situations in a few other places, for instance in `GetResponse` where the 404 for a document not found is actually a valid `GetResponse` rather than an exception. But this case is even weirder, as all the processing happens on the REST layer and the response itself doesn't have a place to store the exception if we wanted to parse it back. I wonder if it would be possible to move this logic to the transport action, and have a lightweight response with much less logic, which holds the final status code and eventual exceptions if there are along with the current content. That would make this request more similar to other cases where for instance we have results together with shard failures that are after all exceptions. What do you think?

Reply:
Do you mean to change the response to something like : 
```
{
  failures : { .... },
  "<index name>" : { ... }, 
  "<index name>" : { ... }
}
```
return 200 and ask the clients to parse the failures to see if there were exceptions ?
Or keep as it currently is and move the filtering logic to the transport?

For moving the logic to the transport, I am not really sure if it is possible. There was an issue and a fix which got reverted because of side effects ( issue : #27763, reverted PR : #28294 ).

Reply:
I mean to leave the final output the same but to refactor the response object. The PR that was reverted was changing things even deeper in the `MetaData` class. I am not sure if this change would end up causing the same problems, I hope not but it may be the case. I see there are subtleties that yield to a different output when comparing the transport client response with what we print out on the REST layer. Maybe then we need to keep what goes in the response exactly the same, still adding room for errors and computing the status in the transport action, but leaving the filtering in the `toXContent`. That should be possible but this is indeed tricky.

Reply:
hey @javanna, sorry I didn't get back on this one sooner.

If the status ( and the error msg ) are calculated on the transport layer, then they shoud also be serialized, shouldn't they ?

Reply:
no worries, yes additional info added as part of the transport action execution should be serialized over transport.